### ISSUE TITLE
[Bug]  upgrade_version: unbound variable

### ISSUE THREAD

**Issue by N-Mi** [Olf0/sfos-upgrade#77]
**SailfishOS VERSION** (Settings → About product → Build): 3.4.0.24
<br />**HARDWARE** (Settings → About product → Manufacturer & Product name): F(x)tec Pro¹ (qx1000, not the Pro¹X which is qx1050)
<br />**sfos-upgrade VERSION** ( `rpm -qi sfos-upgrade` ): 3.9.13
<br />

#### BUG DESCRIPTION

After a fresh reflash of SFOS on a F(x)tec Pro¹ (version 3.4.0.24, which is the latest working image), I tried to upgrade to the latest SFOS version. I downloaded sfos-upgrade 3.9.13 rpm from openrepos (via a browser search).

I tried running sfos-upgrade without arguments, with any supported version as argument, with "--verify", but got this error message :
```
/usr/bin/sfos-upgrade: line 179: upgrade_version: unbound variable
/usr/bin/sfos-upgrade: line 179: installed_version: unbound variable
Comparing versions failed when checking, if downloading lists of SailfishOS releases is necessary.
```

#### STEPS TO REPRODUCE

- flash a new 3.4.0.24 image on F(x)tec Pro¹
- install latest sfos-upgrade release (3.9.13)
- run `sfos-upgrade`

#### ADDITIONAL INFORMATION

```
[root@Pro1 defaultuser]# ssu re
Device release is currently: 3.4.0.24
[root@Pro1 defaultuser]# sfos-upgrade 
/usr/bin/sfos-upgrade: line 179: upgrade_version: unbound variable
/usr/bin/sfos-upgrade: line 179: installed_version: unbound variable
[root@Pro1 defaultuser]# sfos-upgrade --verify
/usr/bin/sfos-upgrade: line 179: upgrade_version: unbound variable
/usr/bin/sfos-upgrade: line 179: installed_version: unbound variable
Comparing versions failed when checking, if downloading lists of SailfishOS releases is necessary.
[root@Pro1 defaultuser]# sfos-upgrade 4.4.0.72
/usr/bin/sfos-upgrade: line 179: upgrade_version: unbound variable
/usr/bin/sfos-upgrade: line 179: installed_version: unbound variable
Comparing versions failed when checking, if downloading lists of SailfishOS releases is necessary.
[root@Pro1 defaultuser]# sfos-upgrade 4.0.1.45
/usr/bin/sfos-upgrade: line 179: upgrade_version: unbound variable
/usr/bin/sfos-upgrade: line 179: installed_version: unbound variable
Comparing versions failed when checking, if downloading lists of SailfishOS releases is necessary.
[root@Pro1 defaultuser]# pkcon get-details sfos-upgrade
Resolving                                                                                                                                                                                                               [                                                                                                           ] (0%)  More than one package matches:
1. sfos-upgrade-3.9.13-release6.noarch [installed]
2. sfos-upgrade-3.9.13-1.2.1.jolla.noarch [sailfishos-chum]
```

Then I installed version 3.9.12 (also from openrepos website), and it worked : 

```
[root@Pro1 defaultuser]# pkcon get-details sfos-upgrade
Resolving                                                                                                                                                                                                               [                                                                                                           ] (0%)  More than one package matches:
1. sfos-upgrade-3.9.12-release5.noarch [installed]
2. sfos-upgrade-3.9.13-1.2.1.jolla.noarch [sailfishos-chum]

#
# at this point, I had made some tests, including a `ssu re 4.4.0.72` and forgot to reset SSU release to 3.4.0.24, which explains why at first it tries to upgrade to 4.4.0.72
#

[root@Pro1 defaultuser]# sfos-upgrade 
Notice: Upgrading from 3.4.0.24 to 4.4.0.72 would omit installing 4.0.1.48 as a stop release!
Warning: Doing so will likely break this SailfishOS installation, unfortunately often only subtly so the issues caused might be observed way later!
Do you really want to continue? (y/N)  
Aborting upon user request.
[root@Pro1 defaultuser]# sfos-upgrade 4.0.1.48
Notice: The installed version 3.4.0.24 is smaller than the one currently set for SSU (4.4.0.72).
A possible reason for this is, that the 'sailfish-osupdateservice osupdate-check' invoked by osupdate-check.service (which is regularly triggered by osupdate-check.timer) does more than just checking (observed with SailfishOS 3.2.1 and earlier): E.g., setting ssu to the recent release version or next stop release.
Never mind, the version for SSU will be set correctly again, later on.

Notice: Mind that sfos-upgrade is best run on a freshly rebooted device.

Notice: Do you want to upgrade SailfishOS from 3.4.0.24 to 4.0.1.48? (y/N) y

Notice: For troubleshooting issues with the upgrade proper, please consult https://jolla.zendesk.com/hc/en-us/articles/360005795474

- Stopping osupdate-check.timer

- Setting SSU to SailfishOS release:
Changing release from 4.4.0.72 to 4.0.1.48
Your device is now in release mode!

- Fetching and installing the SailfishOS upgrade from 3.4.0.24 to 4.0.1.48 (this may take a while):
REFRESHING CACHE AND DOWNLOADING PACKAGES
Finished transaction (status=1, runtime=805987ms)
UPGRADING SYSTEM
Finished transaction (status=1, runtime=273105ms)
FINISHING

```

So it's probably a regression between version 3.9.12 and 3.9.13.
This bug is maybe related to #64

**Comment by Olf0** [comment]
@N-Mi, you wrote:
```
[…]
# at this point, I had made some tests, including a `ssu re 4.4.0.72` and forgot to reset SSU release to 3.4.0.24, which explains why at first it tries to upgrade to 4.4.0.72
#

[root@Pro1 defaultuser]# sfos-upgrade 
Notice: Upgrading from 3.4.0.24 to 4.4.0.72 would omit installing 4.0.1.48 as a stop release!
Warning: Doing so will likely break this SailfishOS installation, unfortunately often only subtly so the issues caused might be observed way later!
Do you really want to continue? (y/N)  
Aborting upon user request.
[root@Pro1 defaultuser]# sfos-upgrade 4.0.1.48
[…]
```

If you use `sfos-upgrade 4.4.0.72` (repetitively, each time followed by a reboot, the final run followed by a `post_sfos-upgrade`), it will guide you through all stop releases as promised in the [README](https://github.com/Olf0/sfos-upgrade#readme).

Only setting a release version by `ssu re <version>` and the executing `sfos-upgrade` without a parameter allows you to "jump over" stop release and to choose a few other dangerous options.

**Comment by Olf0** [comment]
Released [sfos-upgrade 3.9.14](https://github.com/Olf0/sfos-upgrade/releases/tag/3.9.14), which only has this fix added compared to v3.9.13.

**Comment by N-Mi** [comment]
Thanks for your (very) quick fix !

Yes, I tried setting the release using `ssu re` to check if I could find a workaround, but usually I always used the standard way. That was before I found downgrading to 3.9.12 was working.

Anyway, I also found out I messed something when flashing my phone, so I reflashed it directly using a  4.4.0.64 image (which in fact works, as opposed to what I thought).

I need to upgrade now to 4.4.0.72, I'll use the freshly-released 3.9.14 version :)

**Comment by Olf0** [comment]
> Anyway, I also found out I messed something when flashing my phone, so I reflashed it directly using a 4.4.0.64 image (which in fact works, as opposed to what I thought).

That has become Jolla's (especially jovirkku's) mantra in the recent years: Because re-flashing with a current release is so easy, it does not matter that upgrading SailfishOS has become fragile.

> I need to upgrade now to 4.4.0.72, I'll use the freshly-released 3.9.14 version :)

Did all go fine?

BTW, another Jolla-mantra: Only upgrade SailfishOS at the command line if Jolla's GUI updater (in the Settings app) fails (or is not available: on community ports of SailfishOS).

**Comment by N-Mi** [comment]
> Did all go fine?

Yes, I used the RPM generated by Github action (as it was not available in "Assets" at the time I did it.
